azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoaderImpl.java

```diff
@@ -0,0 +1,217 @@
+     * @param dbOperator Interface to perform db related operations.
+     */
+    @Inject
+    public JdbcVersionSetLoaderImpl(DatabaseOperator dbOperator) throws IOException {
+        this.dbOperator = dbOperator;
+        for (VersionSet versionSet : fetchAllVersionSets()) {
+            this.md5ToVersionSet.put(versionSet.getVersionSetMd5Hex(), versionSet);
+            this.numToVersionSet.put(versionSet.getVersionSetNum(), versionSet);
+        }
+    }
+
+    /**
+     * This method inserts the versionSetJsonString and versionSetMd5Hex to the Table version_set
+     * and get the versionSetNum autogenerated post insertion.
+     */
+    @Override
+    public synchronized int insertAndGetVersionSetNum(String versionSetJsonString, String versionSetMd5Hex) throws IOException {
+        final SQLTransaction<Integer> insertVersionSet = transOperator -> {
+            final VersionSetHandler versionSetHandler = new VersionSetHandler();
+            // Check if the versionSet already exists for versionSetMd5Hex
+            List<VersionSet> versionSets = transOperator.query(SELECT_VSET_FROM_MD5, versionSetHandler, versionSetMd5Hex);
+            if (versionSets.size() > 1) {
+                throw new SQLException("Expected only one VersionSet for newly inserted versionSetMd5Hex " + versionSetMd5Hex);
+            }
+            // Insert if it doesn't
+            if (versionSets.isEmpty()) {
+                transOperator.update(INSERT_VSET, versionSetMd5Hex, versionSetJsonString);
+            }
+
+            // Query for versionSet again after insert
+            versionSets = transOperator.query(SELECT_VSET_FROM_MD5, versionSetHandler, versionSetMd5Hex);
+            if (versionSets.size() > 1) {
+                throw new SQLException("Expected only one VersionSet for newly inserted versionSetMd5Hex " + versionSetMd5Hex);
+            }
+            if (versionSets.isEmpty()) {
+                logger.info("No VersionSet exists for newly inserted versionSetMd5Hex " + versionSetMd5Hex);
+                return null;
+            }
+            return versionSets.get(0).getVersionSetNum();
+
+        };
+        try {
+            Integer versionSetNum = this.dbOperator.transaction(insertVersionSet);
+            VersionSet versionSet = new VersionSet(versionSetJsonString, versionSetMd5Hex, versionSetNum);
+            this.md5ToVersionSet.put(versionSetMd5Hex, versionSet);
+            this.numToVersionSet.put(versionSetNum, versionSet);
+            return versionSetNum;
+        } catch (SQLException e) {
+            logger.error("Exception occurred while inserting version set and getting version number", e);
+            throw new IOException("Unable to insert and get versionSetNum", e);
+        }
+    }
+
+    /**
+     * Removes the versionSet row from the table version_set corresponding to versionSetMd5Hex.
+     */
+    @Override
+    public synchronized boolean removeVersionSet(String versionSetMd5Hex) throws IOException {
```

Comment:
version set can be deleted by id as well once it become invalid. Can you please rename this method to deleteVersionSet?

Reply:
Done